sycl/doc/extensions/experimental/sycl_ext_oneapi_inter_process_communication.asciidoc

```diff
@@ -0,0 +1,393 @@
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+[_Note:_ The APIs in this extension uses the concept of a per-platform
+default context as specified in section 4.6.2 "Platform class" of the core SYCL
+specification.
+As a convenience, this extension specification describes the behavior of its
+APIs by using the `khr_get_default_context` function from {khr-default-context}[
+sycl_khr_default_context], however there is no true dependency on that
+extension.
+An implementation could still implement
+sycl_ext_oneapi_inter_process_communication even without implementing
+sycl_khr_default_context because the core SYCL specification still requires
+there to be a per-platform default context even if the core SYCL specification
+does not provide a convenient way to get it.
+_{endnote}_]
+
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension adds the ability for SYCL programs to share device USM memory
+allocations between processes. This is done by the allocating process creating
+a new IPC memory handle through the new free functions and transferring the
+returned handle data to the other processes. The other processes can use the
+handle data to retrieve the corresponding device USM memory.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_INTER_PROCESS_COMMUNICATION` to one of the values defined
+in the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Extension to `enum class aspect`
+
+[source]
+----
+namespace sycl {
+enum class aspect {
+  ...
+  ext_oneapi_ipc_memory
+}
+}
+----
+
+If a SYCL device has this aspect, that device supports the `get` and `open`
+functions specified in the following section.
+
+
+=== Inter-process communicable memory
+
+This extension adds a new type alias `handle_data_t` and a new IPC memory handle
+type `handle` under the `ipc_memory` experimental namespace.
+
+```
+namespace sycl::ext::oneapi::experimental::ipc_memory {
+
+using handle_data_t = sycl::span<const std::byte, dynamic_extent>;
```

Comment:
I feel like it was a mistake to add the "pre-adopted" C++ types like `sycl::span`, and I think it will make our future lives easier if we minimize the things that depend on them.  In general, I'd prefer to rely on `std::span` in cases where we need "span" and then make C++20 a prerequisite for using the API.

In this case, though, is it really that important to use "span"?  I think we expect that the handle will be something like 16 bytes, and I presume the code to get the raw data and transfer it to another process will not be on the hot path.  In any case, the cost to transfer the data to another process will dwarf any savings from "span".  Can we just return a `std::vector`  instead?

Reply:
Right, I feel like we discussed it before. We could add a "view" overload for data function, like for the SYCLBIN extension.

Reply:
It's indeed a very small piece of memory, but a potential heap allocation nonetheless. Adding the view interfaces don't necessarily add a lot of complexity to the implementation, but adds functionality that the user cannot add themselves. I personally think it is worth the effort, even if it is just crumbs.